File: packages/flutter/test/language/domain/value_object/language_type_value_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:allowance_questboard/language/enum/value_object/language_type_value.dart';
import 'package:allowance_questboard/language/enum/value_object/language_id.dart';
import 'package:allowance_questboard/language/enum/value_object/language_code.dart';
import 'package:allowance_questboard/language/enum/value_object/language_name.dart';
import 'package:allowance_questboard/login/api/v1/init/init_api_response.dart' show LanguageDto;
import 'package:allowance_questboard/shared/domain/value_object/sort_order.dart';

void main() {
  group('LanguageTypeValue', () {
    group('constructor', () {
      test('必須のIDを指定して初期化できること', () {
        // 準備
        final id = LanguageId(1);
        
        // 実行
        final languageTypeValue = LanguageTypeValue(id: id);
        
        // 検証
        expect(languageTypeValue.id, equals(id));
        expect(languageTypeValue.languageId, equals(id));
        expect(languageTypeValue.code.value, equals("0"));
        expect(languageTypeValue.name.value, equals("Unknown"));
        expect(languageTypeValue.isActive, equals(false));
        expect(languageTypeValue.sortOrder.value, equals(0));
      });

      test('全ての値を指定して初期化できること', () {
        // 準備
        final id = LanguageId(1);
        final code = LanguageCode("ja");
        final name = LanguageName("日本語");
        final sortOrder = SortOrder(1);
        
        // 実行
        final languageTypeValue = LanguageTypeValue(
          id: id,
          code: code,
          name: name,
          isActive: true,
          sortOrder: sortOrder,
        );
        
        // 検証
        expect(languageTypeValue.id, equals(id));
        expect(languageTypeValue.languageId, equals(id));
        expect(languageTypeValue.code, equals(code));
        expect(languageTypeValue.name, equals(name));
        expect(languageTypeValue.isActive, equals(true));
        expect(languageTypeValue.sortOrder, equals(sortOrder));
      });
    });

    group('setFromDto', () {
      test('DTOから値を設定できること', () {
        // 準備
        final id = LanguageId(1);
        final languageTypeValue = LanguageTypeValue(id: id);
        final dto = LanguageDto(
          id: 1,
          code: "en",
          name: "English",
          isActive: true,
          sortOrder: 2,
        );
        
        // 実行
        languageTypeValue.setFromDto(dto);
        
        // 検証
        expect(languageTypeValue.code.value, equals("en"));
        expect(languageTypeValue.name.value, equals("English"));
        expect(languageTypeValue.isActive, equals(true));
        expect(languageTypeValue.sortOrder.value, equals(2));
      });

      test('DTOから値を設定してもIDは変更されないこと', () {
        // 準備
        final id = LanguageId(5);
        final languageTypeValue = LanguageTypeValue(id: id);
        final dto = LanguageDto(
          id: 1,
          code: "en",
          name: "English",
          isActive: true,
          sortOrder: 2,
        );
        
        // 実行
        languageTypeValue.setFromDto(dto);
        
        // 検証
        expect(languageTypeValue.id, equals(id));
        expect(languageTypeValue.languageId, equals(id));
      });
    });
  });
}

File: packages/backend/__tests__/language/value-object/languageTypeValue.test.ts
import { describe, test, expect } from '@jest/globals';
import { LanguageTypeValue, LanguageTypeValueSchema } from '@shared/features/language/value-object/languageTypeValue';
import { LanguageId } from '@shared/features/language/value-object/languageId';
import { LanguageCode } from '@shared/features/language/value-object/languageCode';
import { LanguageName } from '@shared/features/language/value-object/languageName';
import { LanguageEntity } from '@backend/features/language/entity/languageEntity';

describe('LanguageTypeValue', () => {
  describe('コンストラクタでインスタンスを作成すること', () => {
    test('デフォルト値でインスタンスが作成されること', () => {
      // 準備
      const id = new LanguageId(1);
      
      // 実行
      const languageTypeValue = new LanguageTypeValue(id);
      
      // 検証
      expect(languageTypeValue.id).toBe(id);
      expect(languageTypeValue.code.value).toBe('unknown');
      expect(languageTypeValue.name.value).toBe('Unknown');
      expect(languageTypeValue.isActive).toBe(false);
      expect(languageTypeValue.sortOrder).toBe(0);
    });

    test('すべてのパラメータを指定してインスタンスが作成されること', () => {
      // 準備
      const id = new LanguageId(1);
      const code = new LanguageCode('ja');
      const name = new LanguageName('Japanese');
      const isActive = true;
      const sortOrder = 1;
      
      // 実行
      const languageTypeValue = new LanguageTypeValue(id, code, name, isActive, sortOrder);
      
      // 検証
      expect(languageTypeValue.id).toBe(id);
      expect(languageTypeValue.code).toBe(code);
      expect(languageTypeValue.name).toBe(name);
      expect(languageTypeValue.isActive).toBe(isActive);
      expect(languageTypeValue.sortOrder).toBe(sortOrder);
    });
  });

  describe('setFromEntityメソッドでエンティティから値を設定すること', () => {
    test('エンティティの値が正しく設定されること', () => {
      // 準備
      const id = new LanguageId(1);
      const languageTypeValue = new LanguageTypeValue(id);
      const entity = Object.assign(new LanguageEntity(), {
        id: 1,
        code: 'en',
        name: 'English',
        is_active: true,
        sort_order: 2
      });
      
      // 実行
      languageTypeValue.setFromEntity(entity);
      
      // 検証
      expect(languageTypeValue.code.value).toBe('en');
      expect(languageTypeValue.name.value).toBe('English');
      expect(languageTypeValue.isActive).toBe(true);
      expect(languageTypeValue.sortOrder).toBe(2);
    });
  });

  describe('toZodDataメソッドでZodスキーマに準拠したデータを返すこと', () => {
    test('正しいデータ構造が返されること', () => {
      // 準備
      const id = new LanguageId(1);
      const code = new LanguageCode('ja');
      const name = new LanguageName('Japanese');
      const languageTypeValue = new LanguageTypeValue(id, code, name, true, 1);
      
      // 実行
      const zodData = languageTypeValue.toZodData();
      
      // 検証
      expect(zodData.id).toBe(1);
      expect(zodData.code).toBe('ja');
      expect(zodData.name).toBe('Japanese');
      expect(zodData.isActive).toBe(true);
      expect(zodData.sortOrder).toBe(1);
    });

    test('Zodスキーマによる検証が成功すること', () => {
      // 準備
      const id = new LanguageId(1);
      const languageTypeValue = new LanguageTypeValue(id);
      
      // 実行
      const zodData = languageTypeValue.toZodData();
      const validationResult = LanguageTypeValueSchema.safeParse(zodData);
      
      // 検証
      expect(validationResult.success).toBe(true);
    });
  });

  describe('setFromZodDataメソッドでZodデータから値を初期化すること', () => {
    test('正しいデータで値が設定されること', () => {
      // 準備
      const id = new LanguageId(1);
      const languageTypeValue = new LanguageTypeValue(id);
      const zodData = {
        id: 1,
        code: 'en',
        name: 'English',
        isActive: true,
        sortOrder: 2
      };
      
      // 実行
      languageTypeValue.setFromZodData(zodData);
      
      // 検証
      expect(languageTypeValue.code.value).toBe('en');
      expect(languageTypeValue.name.value).toBe('English');
      expect(languageTypeValue.isActive).toBe(true);
      expect(languageTypeValue.sortOrder).toBe(2);
    });

    test('一部のプロパティが未定義でもデフォルト値が設定されること', () => {
      // 準備
      const id = new LanguageId(1);
      const languageTypeValue = new LanguageTypeValue(id);
      const zodData = {
        id: 1
      };
      
      // 実行
      languageTypeValue.setFromZodData(zodData);
      
      // 検証
      expect(languageTypeValue.code.value).toBe('unknown');
      expect(languageTypeValue.name.value).toBe('Unknown');
      expect(languageTypeValue.isActive).toBe(false);
      expect(languageTypeValue.sortOrder).toBe(0);
    });

    test('IDが異なる場合にエラーが発生すること', () => {
      // 準備
      const id = new LanguageId(1);
      const languageTypeValue = new LanguageTypeValue(id);
      const zodData = {
        id: 2,
        code: 'en',
        name: 'English'
      };
      
      // 実行・検証
      expect(() => {
        languageTypeValue.setFromZodData(zodData);
      }).toThrow('ID mismatch: expected 1, got 2');
    });
  });
});
